lib/symboldatabase.h

```diff
@@ -1151,7 +1151,8 @@ class CPPCHECKLIB ValueType {
     enum Sign { UNKNOWN_SIGN, SIGNED, UNSIGNED } sign;
     enum Type { UNKNOWN_TYPE, NONSTD, RECORD, CONTAINER, ITERATOR, VOID, BOOL, CHAR, SHORT, WCHAR_T, INT, LONG, LONGLONG, UNKNOWN_INT, FLOAT, DOUBLE, LONGDOUBLE } type;
     nonneg int bits;                    ///< bitfield bitcount
-    nonneg int pointer;                 ///< 0=>not pointer, 1=>*, 2=>**, 3=>***, etc
+    nonneg int pointerDepth;            ///< 0=>not pointer, 1=>*, 2=>**, 3=>***, etc
+    bool reference;                     ///< Is the outermost pointer actually a reference. i.e. if pointer==2 then are we dealing with an X** or an X*&
```

Comment:
hmm.. how about `X&`  => pointerDepth=0, reference=true
I am not sure how that will affect all the code in practice I believe you have a better idea. It just seems a bit more intuitive? To calculate the type for a `*p` expression we always decrease `pointerDepth` with 1 and clear `reference`.

Reply:
I think that's a very good suggestion. I think it will allow me to limit the scope of my changes so that we can add the knowledge of references incrementally rather than being forced to update everything. Of course that probably means we'll miss a bunch of existing bugs where references aren't handled properly but it's a lot less work and much less likely to introduce new bugs.